[DRAFT] Build: Refactor building tool to oxc for performance

DieterReinert opened the pull request:
### Summary
Replace the existing build tool with OXC for significant build performance improvements.

### Changes
- Replace the JavaScript/TypeScript transformation tool with oxc-transform.
- Update tools/build-utils.js to use OXC's transform API.
- Update the build script to include oxc-transform in dependency checks.
- Update package dependencies and lock file.

### Motivation
OXC is a Rust-based JavaScript/TypeScript parser and transformer that offers significantly faster build times compared to ESBuild.

### Performance Impact
[TODO: Add build time benchmarks comparing before/after]

### To-do's
- Benchmarks
- Replace tsc with unplugin-isolated-decl for even more performance.
- Update GitHub CI/CD
- Wait for OXC Transformer to support cjs (relying on surcase)
- Benchmark native Node's native typescript support and check hotpatch

DieterReinert commented:
This is a draft as OXC Transformer is still in Alpha. @Zarel

mia-pi-git commented:
> This is a draft as OXC Transformer is still in Alpha. @Zarel

I don’t really think this is necessary? ESBuild works well and is speedy for the tasks we need, adding a bunch of extra dependencies (and one in alpha, no less) doesn’t seem like the best use of time and potential maintenance burden for a minimal speedup.

DieterReinert commented:
> > This is a draft as OXC Transformer is still in Alpha. @Zarel
> 
> I don’t really think this is necessary? ESBuild works well and is speedy for the tasks we need, adding a bunch of extra dependencies (and one in alpha, no less) doesn’t seem like the best use of time and potential maintenance burden for a minimal speedup.

I discussed this with Zarel and we came to the conclusion to set it as draft until it's ready. The state of the package is indeed alpha and that's why the pr is still a draft.

The surcase dependency should disappear when oxc-transformer adds common js support.

I had tested it on my pc and my builds went from 5 to 1 second and there are still some optimizations left.

Zarel commented on build:
```diff
@@ -21,14 +21,13 @@ function shell(cmd) {
-	var version = require('esbuild').version.split('.');
-	if (parseInt(version[1]) < 16) {
-		throw new Error("esbuild version too old");
+for (var pkg of ['oxc-transform', 'fast-glob', 'typescript', 'sucrase']) {
+	try {
+		require.resolve(pkg);
+	} catch (e) {
+		console.log('Installing dependencies...');
+		shell('npm install ' + pkg);
```
We usually just put the most recent added/updated package here, instead of needing to loop through multiple of them.

Zarel commented on build:
```diff
@@ -39,14 +38,13 @@ try {
 	if (err.code !== 'MODULE_NOT_FOUND') throw err; // should never happen
 
 	console.log('config.js does not exist. Creating one with default settings...');
+	fs.mkdirSync('config', {recursive: true});
```
...AI code?

My main concern is that it'll clutter up the diff with useless changes like this one.

```suggestion
```

Zarel commented on tools/build-utils.js:
```diff
@@ -1,66 +1,74 @@
-			fs.copyFileSync(`${file}/${f}`, require('path').resolve('dist', `${file}/${f}`));
-		}
+function copyJSON(dir = 'data') {
+	for (const src of fg.sync(`${dir}/**/*.json`)) {
```
I would be very surprised if this made enough of a difference to be worth grabbing the fast-glob dependency...

Slayer95 commented:
Node 18 is EOL, and already has known unpatched vulnerabilities. The Node.js Foundation recommends upgrading to Node v22 
https://nodejs.org/en/blog/announcements/node-18-eol-support

DieterReinert commented:
I made a quick sketch of what a glob function would look like:
```JS
function customGlobSync(patterns) {
	const patternArray = Array.isArray(patterns) ? patterns : [patterns];

	const includePatterns = [];
	const excludePatterns = [];

	for (const pattern of patternArray) {
		if (pattern.startsWith('!')) {
			excludePatterns.push(pattern.slice(1));
		} else {
			includePatterns.push(pattern);
		}
	}

	function globToRegex(glob) {
		let pattern = glob;

		pattern = pattern.replace(/\\/g, '/');

		pattern = pattern.replace(/[.+^$()[\]\\|]/g, '\\$&');

		pattern = pattern.replace(/\{([^}]+)\}/g, (match, content) => {
			const options = content.split(',').map(s => s.trim());
			return '(' + options.join('|') + ')';
		});

		pattern = pattern.replace(/\*\*/g, '___DOUBLE_STAR___');

		pattern = pattern.replace(/\*/g, '[^/]*');

		pattern = pattern.replace(/\?/g, '[^/]');

		pattern = pattern.replace(/^___DOUBLE_STAR___\//g, '(.*/)?');
		pattern = pattern.replace(/\/___DOUBLE_STAR___\//g, '(/.*)?/');
		pattern = pattern.replace(/\/___DOUBLE_STAR___$/g, '(/.*)?');
		pattern = pattern.replace(/___DOUBLE_STAR___/g, '.*');

		return new RegExp('^' + pattern + '$');
	}

	function matchesAnyPattern(filePath, patterns) {
		if (patterns.length === 0) return false;

		const normalizedPath = filePath.replace(/\\/g, '/');
		return patterns.some(pattern => {
			const regex = globToRegex(pattern);
			return regex.test(normalizedPath);
		});
	}

	function walkDirectory(dir, baseDir = dir) {
		const results = [];

		try {
			const entries = fs.readdirSync(dir, {withFileTypes: true});

			for (const entry of entries) {
				const fullPath = path.join(dir, entry.name);
				const relativePath = path.relative(baseDir, fullPath).replace(/\\/g, '/');

				if (excludePatterns.length > 0 && matchesAnyPattern(relativePath, excludePatterns)) {
					continue;
				}

				if (entry.isDirectory()) {
					let shouldRecurse = false;

					if (includePatterns.length === 0) {
						shouldRecurse = true;
					} else {
						for (const pattern of includePatterns) {
							if (!pattern.includes('/')) {
								continue;
							}
							if (pattern.includes('**')) {
								shouldRecurse = true;
								break;
							}
							const dirPrefix = relativePath + '/';
							if (pattern.startsWith(dirPrefix)) {
								shouldRecurse = true;
								break;
							}
							const parts = pattern.split('/');
							const dirParts = relativePath.split('/');
							if (parts.length > dirParts.length) {
								let matches = true;
								for (let i = 0; i < dirParts.length; i++) {
									const patternPart = parts[i];
									const dirPart = dirParts[i];
									if (patternPart !== dirPart &&
										!patternPart.includes('*') &&
										!patternPart.includes('?') &&
										!patternPart.includes('{')) {
										matches = false;
										break;
									}
								}
								if (matches) {
									shouldRecurse = true;
									break;
								}
							}
						}
					}

					if (shouldRecurse) {
						results.push(...walkDirectory(fullPath, baseDir));
					}
				} else if (entry.isFile()) {
					if (includePatterns.length === 0 || matchesAnyPattern(relativePath, includePatterns)) {
						results.push(relativePath);
					}
				}
			}
		} catch (err) {
			if (err.code !== 'EACCES' && err.code !== 'EPERM') {
				console.error(`Error reading directory ${dir}:`, err);
			}
		}

		return results;
	}

	return walkDirectory('.', '.');
}
```

I'm not sure if that's something we want to maintain.
